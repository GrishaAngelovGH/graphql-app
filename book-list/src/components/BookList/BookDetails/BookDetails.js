import ListGroup from "react-bootstrap/ListGroup"

const BookDetails = () => {
  return (
    <div className="row g-0 justify-content-center">
      <div className="col-md-11">
        <h3 className="text-center text-secondary fw-bold">Book Details</h3>

        <ListGroup>
          <ListGroup.Item>
            <span className="fw-bold mx-2">Title:</span>
            <span></span>
          </ListGroup.Item>
          <ListGroup.Item>
            <span className="fw-bold mx-2">Author:</span>
            <span></span>
          </ListGroup.Item>
        </ListGroup>

        <ListGroup className="mt-3">
          <ListGroup.Item className="text-center">
            <span className="fw-bold">Author's Books</span>
          </ListGroup.Item>
          <div className="overflow-auto" style={{ height: 'calc(100vh - 300px)' }}>
            <ListGroup.Item></ListGroup.Item>
          </div>
        </ListGroup>
      </div>
    </div>
  )
}

export default BookDetails