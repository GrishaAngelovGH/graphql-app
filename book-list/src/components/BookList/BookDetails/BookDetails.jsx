import { useState } from "react"
import { useMutation, useQuery } from "@apollo/client"

import Button from "react-bootstrap/Button"
import ListGroup from "react-bootstrap/ListGroup"

import ConfirmationModal from "components/ConfirmationModal"

import { DELETE_BOOK } from "mutations/apollo-mutations"
import { GET_BOOKS, GET_BOOK_BY_ISBN } from "queries/apollo-queries"

const BookDetails = ({ isbn, onAfterDeleteBook }) => {
  const [showConfirm, setShowConfirm] = useState(false)

  const { loading, data } = useQuery(GET_BOOK_BY_ISBN, {
    variables: { isbn }
  })

  const [deleteBook] = useMutation(DELETE_BOOK)

  const handleShowConfirm = () => {
    setShowConfirm(!showConfirm)
  }

  const handleConfirmAction = () => {
    setShowConfirm(!showConfirm)

    deleteBook({
      variables: {
        isbn,
        authorId: data.book.author.id
      },
      refetchQueries: [
        { query: GET_BOOKS }
      ]
    })

    onAfterDeleteBook()
  }

  return (
    <div className="row g-0 justify-content-center">
      <div className="col-md-11">
        <h3 className="text-center text-secondary fw-bold">Book Details</h3>

        {loading && <p className="text-center">Loading...</p>}

        <ListGroup>
          <ListGroup.Item>
            <span className="fw-bold mx-2">Title:</span>
            <span>{data?.book.title}</span>
          </ListGroup.Item>
          <ListGroup.Item>
            <span className="fw-bold mx-2">Author:</span>
            <span>{data?.book.author.name}</span>
          </ListGroup.Item>
          <ListGroup.Item className="text-center">
            <Button
              variant="danger"
              className="w-50"
              onClick={handleShowConfirm}
            >
              Delete Book
            </Button>
          </ListGroup.Item>
        </ListGroup>

        <ListGroup className="mt-3">
          <ListGroup.Item className="text-center">
            <span className="fw-bold">Author&apos;s Books</span>
          </ListGroup.Item>
          <div className="overflow-auto" style={{ height: 'calc(100vh - 350px)' }}>
            {
              data?.book.author.books.map(v => (
                <ListGroup.Item key={v.isbn}>{v.title}</ListGroup.Item>
              ))
            }
          </div>
        </ListGroup>

        <ConfirmationModal
          show={showConfirm}
          title="Confirm Delete"
          body="Are you sure that you want to delete this book?"
          onClose={handleShowConfirm}
          onConfirm={handleConfirmAction}
        />
      </div>
    </div>
  )
}

export default BookDetails