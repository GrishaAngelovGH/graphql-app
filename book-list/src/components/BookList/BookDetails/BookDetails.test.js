import { render } from '@testing-library/react'

import BookDetails from './BookDetails'

test('should render BookDetails component', () => {
  const view = render(<BookDetails />)

  expect(view).toMatchSnapshot()
})