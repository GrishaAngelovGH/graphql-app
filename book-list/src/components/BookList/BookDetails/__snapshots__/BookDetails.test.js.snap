// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BookDetails component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="row g-0 justify-content-center"
      >
        <div
          class="col-md-11"
        >
          <h3
            class="text-center text-secondary fw-bold"
          >
            Book Details
          </h3>
          <div
            class="list-group"
          >
            <div
              class="list-group-item"
            >
              <span
                class="fw-bold mx-2"
              >
                Title:
              </span>
              <span>
                Title 1
              </span>
            </div>
            <div
              class="list-group-item"
            >
              <span
                class="fw-bold mx-2"
              >
                Author:
              </span>
              <span>
                Author 1
              </span>
            </div>
            <div
              class="text-center list-group-item"
            >
              <button
                class="w-50 btn btn-danger"
                type="button"
              >
                Delete Book
              </button>
            </div>
          </div>
          <div
            class="mt-3 list-group"
          >
            <div
              class="text-center list-group-item"
            >
              <span
                class="fw-bold"
              >
                Author's Books
              </span>
            </div>
            <div
              class="overflow-auto"
              style="height: calc(100vh - 350px);"
            >
              <div
                class="list-group-item"
              >
                Title 1
              </div>
              <div
                class="list-group-item"
              >
                Title 2
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="row g-0 justify-content-center"
    >
      <div
        class="col-md-11"
      >
        <h3
          class="text-center text-secondary fw-bold"
        >
          Book Details
        </h3>
        <div
          class="list-group"
        >
          <div
            class="list-group-item"
          >
            <span
              class="fw-bold mx-2"
            >
              Title:
            </span>
            <span>
              Title 1
            </span>
          </div>
          <div
            class="list-group-item"
          >
            <span
              class="fw-bold mx-2"
            >
              Author:
            </span>
            <span>
              Author 1
            </span>
          </div>
          <div
            class="text-center list-group-item"
          >
            <button
              class="w-50 btn btn-danger"
              type="button"
            >
              Delete Book
            </button>
          </div>
        </div>
        <div
          class="mt-3 list-group"
        >
          <div
            class="text-center list-group-item"
          >
            <span
              class="fw-bold"
            >
              Author's Books
            </span>
          </div>
          <div
            class="overflow-auto"
            style="height: calc(100vh - 350px);"
          >
            <div
              class="list-group-item"
            >
              Title 1
            </div>
            <div
              class="list-group-item"
            >
              Title 2
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
